### PR TITLE
cmd/snap-update-ns: keep order of existing mount entries

### PR DESCRIPTION
**Pull request by zyga**
This is stacked on top of https://github.com/canonical/snapd/pull/14491

We recently realized that the order of saved mount profile, after
performing mount namespace update, was subtly incorrect. The order was
seemingly swapped after each operation. In reality, the order reflected
the order of performed changes, including the special change type of
"keep", which means that an entry is kept intact.

Fix the problem by storing "kept" changes in the original order, which
is back-to-front compared to the order of changes computed by the
algorithm. At the same time, keep track of mount changes in the order of
their execution.

This probably fixes a hell lot of edge cases and weird behaviors that
went undiagnosed for many years. Interestingly a test case had a TODO
that hinted at the problem, but it was not clear what the problem was at
the time. The TODO is now gone and the test case is now up-to-date.

Test data for cmd/snap-update-ns/testdata was re-generated.

Jira: https://warthogs.atlassian.net/browse/SNAPDENG-31644

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: cmd/snap-update-ns/update.go [hunk 1]
```diff
@@ -93,8 +93,14 @@ func executeMountProfileUpdate(upCtx MountProfileUpdateContext) error {
 	// Compute the new current profile so that it contains only changes that were made
 	// and save it back for next runs.
 	var currentAfter osutil.MountProfile
+	for i := len(changesMade) - 1; i >= 0; i-- {
```
Comment (bboozzoo) [inline cmd/snap-update-ns/update.go:96]: I think it's useful to leave a comment for ourselves that the initial state was initially iterated over from back to front, so the things we keep are also present in reverse order. Thus order to preserve the stable state for kept entries we should record them in reverse order.

File: cmd/snap-update-ns/change.go [hunk 2]
```diff
@@ -65,6 +65,10 @@ type Change struct {
 	Action Action
 }
 
+func (c *Change) GoString() string {
```
[the remark this answers is not in the document]
Comment (zyga) [reply cmd/snap-update-ns/change.go:68]: It's useful for printing changes. It's not use here but really valuable in practice.

File: cmd/snap-update-ns/update.go (hunk 1 above)
Comment (zyga) [reply cmd/snap-update-ns/update.go]: Pushed now

File: cmd/snap-update-ns/update.go [hunk 3]
```diff
@@ -92,11 +92,24 @@ func executeMountProfileUpdate(upCtx MountProfileUpdateContext) error {
...
 	// and save it back for next runs.
-	var currentAfter osutil.MountProfile
-	for _, change := range changesMade {
-		if change.Action == Mount || change.Action == Keep {
-			currentAfter.Entries = append(currentAfter.Entries, change.Entry)
+	currentAfter := CurrentProfileFromChangesMade(changesMade)
+	return upCtx.SaveCurrentProfile(&currentAfter)
+}
+
+func CurrentProfileFromChangesMade(changes []*Change) osutil.MountProfile {
```
Comment (pedronis) [inline cmd/snap-update-ns/update.go:99]: this needs a comment of its own

**Review by pedronis — APPROVED** [review]
as I said looking at how NeededChanges works at the moment it appears this will mostly affect the order of unmounting of things we kept that we cannot keep anymore the next time around, it looks the right direction but it's hard to predict the general effects, as the logic in NeededChanges is helpful to an extent but as we know has limitations and is hard to predict atm

File: cmd/snap-update-ns/update.go (hunk 3 above)
Comment (zyga) [reply cmd/snap-update-ns/update.go:115]: Done

**Review by bboozzoo — APPROVED** [review]
Thanks
